Spigot-API-Patches/0018-Add-exception-reporting-event.patch

```diff
@@ -457,6 +457,31 @@ index 00000000..5582999f
 +        super(message, cause, enableSuppression, writableStackTrace, command, commandSender, arguments);
 +    }
 +}
+diff --git a/src/main/java/org/bukkit/command/PluginCommand.java b/src/main/java/org/bukkit/command/PluginCommand.java
+index 2abe1208..d97c66c4 100644
+--- a/src/main/java/org/bukkit/command/PluginCommand.java
++++ b/src/main/java/org/bukkit/command/PluginCommand.java
+@@ -5,6 +5,12 @@ import java.util.List;
+ import org.apache.commons.lang.Validate;
+ import org.bukkit.plugin.Plugin;
+ 
++// Paper
++import org.bukkit.Bukkit;
++import com.destroystokyo.paper.event.server.ServerExceptionEvent;
++import com.destroystokyo.paper.exception.ServerCommandException;
++// Paper end
++
+ /**
+  * Represents a {@link Command} belonging to a plugin
+  */
+@@ -43,6 +49,7 @@ public final class PluginCommand extends Command implements PluginIdentifiableCo
+         try {
+             success = executor.onCommand(sender, this, commandLabel, args);
+         } catch (Throwable ex) {
++            Bukkit.getPluginManager().callEvent(new ServerExceptionEvent(new ServerCommandException(ex, this, sender, args))); // Paper
```

Comment:
I didn't entirely clarify in my comment too well, but I personally feel that this should be inside of SimpleCommandMap, especially if the intent is to ensure that we capture and throw an event for all exceptions from commands; any plugin using a command framework which registers its own commands and uses the CommandException properly as expected would not call this event

Reply:
Maybe instead of inside SimpleCommandMap, its called at the caller to SCM .onCommand instead, for cases where hacky plugins implement custom command maps.

Reply:
But yeah as @electronicboy said, its very common for plugins to not even use PluginCommand, so this won't be sufficient.